app/src/main/java/org/simple/clinic/patient/PatientFuzzySearch.kt

```diff
@@ -0,0 +1,160 @@
+import android.arch.persistence.db.SupportSQLiteOpenHelper
+import android.database.Cursor
+import io.reactivex.Completable
+import io.reactivex.Single
+import io.reactivex.SingleTransformer
+import java.util.UUID
+
+class PatientFuzzySearch {
+
+  companion object {
+
+    // TODO: I am really unhappy about putting this here, but I can't figure out where else to place it.
+    //
+    // This needs to be accessible both from the AppDatabase builder (since we can't create this table
+    // using the Room classes and requires running a specific SQL query) as well as the migration.
+    //
+    // At the same time, this is very specific to the database and should not be exposed to the rest
+    // of the system as a top-level function.
+    @JvmStatic
+    fun createTable(database: SupportSQLiteDatabase) {
+      database.execSQL("""CREATE VIRTUAL TABLE "PatientFuzzySearch" USING spellfix1""")
+    }
+
+    @JvmStatic
+    fun clearTable(database: SupportSQLiteDatabase) {
+      database.execSQL("""DELETE FROM "PatientFuzzySearch"""")
+    }
+  }
+
+  data class FuzzySearchResult(
+      val rowId: Long,
+      val uuid: UUID,
+      val word: String,
+      val rank: Int,
+      val distance: Int,
+      val score: Int,
+      val matchLength: Int,
+      val scope: Int
+  )
+
+  data class UuidToScore(val uuid: UUID, val score: Int)
+
+  interface PatientFuzzySearchDao {
+
+    // Used for testing
+    fun savedEntries(): Single<List<FuzzySearchResult>>
+
+    // Used for testing
+    fun getEntriesForPatientIds(uuids: List<UUID>): Single<List<FuzzySearchResult>>
+
+    fun updateTableForPatients(uuids: List<UUID>): Completable
+
+    fun searchForPatientsWithNameLike(query: String): Single<List<PatientSearchResult>>
+
+    fun searchForPatientsWithNameLikeAndAgeWithin(query: String, dobUpperBound: String, dobLowerBound: String): Single<List<PatientSearchResult>>
+  }
+
+  class PatientFuzzySearchDaoImpl(
+      private val sqLiteOpenHelper: SupportSQLiteOpenHelper,
+      private val patientSearchDao: PatientSearchResult.RoomDao
+  ) : PatientFuzzySearchDao {
+
+    private val fuzzyColumnsToFetch = """
+      "P"."uuid" "uuid", "PFS"."rowid" "rowid","PFS"."word" "word","PFS"."rank" "rank","PFS"."distance" "distance","PFS"."score" "score","PFS"."matchlen" "matchlen","PFS"."scope" "scope"
+    """.trimIndent()
+
+    private fun resultsFromCursor(cursor: Cursor) =
+        cursor.use {
+          val uuid = it.getColumnIndex("uuid")
+          val rowId = it.getColumnIndex("rowid")
+          val word = it.getColumnIndex("word")
+          val rank = it.getColumnIndex("rank")
+          val distance = it.getColumnIndex("distance")
+          val score = it.getColumnIndex("score")
+          val matchLength = it.getColumnIndex("matchlen")
+          val scope = it.getColumnIndex("scope")
+          generateSequence { it.takeIf { it.moveToNext() } }
+              .map {
+                FuzzySearchResult(
```

Comment:
Hey, this code looks new. Do you think we can move this mapper function to the model? This idea is from `Sqlbrite`, where the model implements a mapping function and you just call `FuzzySearchResult.mapFrom(cursor)` or something.

Reply:
Noted.